### PR TITLE
`AsyncStop` option for graceful stop in async mode

### PR DESCRIPTION
**Pull request by JamesJJ**
As I understand, existing behaviour in Async mode is:

 * Add events to the end of the buffer. Return an error if the buffer is full
 * Take events from the front of the buffer, for each event try to send to fluentd `MaxRetry` times, and if still not successfully sent the event is discarded
 * When `Close()` is called, the pending event channel is closed and attempting to add any new events will return an error
 * When `Close()` is called, it will block until all remaining events in the buffer have either been successfully sent, or each one has run through it's `MaxRetry` number of failures. This means that `Close()` can block for a very long time if the number of unsent events is large.

This PR introduces a new option `AsyncStop`. When true in Async mode and Close() is called, the current write() / MaxRetries iteration will complete for the single event in-progress, and then any remaining events in the buffer will be discarded. This may help to significantly reduce the time that `Close()` blocks when fluentd is continuously available.

 * Possible fix for #71

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @tagomoris @cpuguy83 ptal (looks like this relates to https://github.com/moby/moby/issues/40063)

**Comment by cpuguy83** [comment]
In theory this is nice, but I feel like everyone will just turn it on because close just should not block in general.
The option also doesn't seem appropriately named since it's not an async close but rather discard+close.

Perhaps something like `Reset()` before `Close()` would be a possible alternative.
But again I feel like probably everyone will just want to call `Reset()`.

File: fluent/fluent.go [hunk 1]
```diff
@@ -347,13 +353,25 @@ func (f *Fluent) connect() (err error) {
...
 
 func (f *Fluent) run() {
+	drainEvents := false
 	for {
+		select {
+		case stopRunning, ok := <-f.stopRunning:
+			if stopRunning || !ok {
+				drainEvents = true
+			}
+default:
```
Comment (tagomoris) [inline fluent/fluent.go]: Could you use non-`default` label here (and specify it on `continue`)?
This is a part of the entire loop, and going back to here is a kind of exceptional case. So it should be not `default`.

File: fluent/fluent.go [hunk 2]
```diff
@@ -49,6 +49,7 @@ type Config struct {
 	MaxRetryWait     int           `json:"max_retry_wait"`
 	TagPrefix        string        `json:"tag_prefix"`
 	Async            bool          `json:"async"`
+	AsyncStop        bool          `json:"async_stop"`
```
Comment (tagomoris) [inline fluent/fluent.go]: `AsyncStop` seems something to specify to "stop eventually" to me, and misses to explain that events will be discarded.
How about using `ForceStopAsyncSend` or something else?
(I don't think that name is good enough. Different ideas are welcome :D)

File: fluent/fluent.go [hunk 3]
```diff
@@ -347,13 +353,25 @@ func (f *Fluent) connect() (err error) {
...
+default:
+		}
 		select {
 		case entry, ok := <-f.pending:
 			if !ok {
 				f.wg.Done()
 				return
 			}
+			if drainEvents {
+				fmt.Fprintf(os.Stderr, "[%s] Unable to send logs to fluentd, discarding...\n", time.Now().Format(time.RFC3339))
```
Comment (tagomoris) [inline fluent/fluent.go]: This feature is for the case when the destination can't receive events, so the buffer should contain so many entries. If the buffer has many entries to be sent, this message will be shown so many times (and log flood will happen).
Can you update the code to show this message just once (or few times)?

Comment (JamesJJ) [reply fluent/fluent.go]: Yes, no problem

File: fluent/fluent.go [hunk 4]
```diff
@@ -347,13 +353,26 @@ func (f *Fluent) connect() (err error) {
 }
 
 func (f *Fluent) run() {
+	drainEvents := false
+	var emitEventDrainMsg sync.Once
 	for {
+		select {
```
Comment (tagomoris) [inline fluent/fluent.go]: I missed that `default:` is for this `select` clause...
But if so, this `select` will be processed every time in the very busy loop to drain events from `f.pending`.
Can we skip it `if drainEvents`?

**Comment by tagomoris** [comment]
LGTM. Thank you!

**Comment by tagomoris** [comment]
Merged, and tagged as v1.5.0

**Pull request merged**